packages/vega-scenegraph/src/HybridRenderer.js

```diff
@@ -0,0 +1,96 @@
+import Renderer from './Renderer';
+import SVGRenderer from './SVGRenderer';
+import CanvasRenderer from './CanvasRenderer';
+import {inherits} from 'vega-util';
+import {domChild} from './util/dom';
+
+const OPTS = {svgMarkTypes: ['text']};
+
+export function setHybridRendererOptions(svgMarkTypes, debug) {
+  OPTS['svgMarkTypes'] = svgMarkTypes;
+  OPTS['debug'] = debug ?? false;
+}
+
+export default function HybridRenderer(loader) {
+  Renderer.call(this, loader);
+  this._svgRenderer = new SVGRenderer(loader);
+  this._canvasRenderer = new CanvasRenderer(loader);
+}
+
+const base = Renderer.prototype;
+
+inherits(HybridRenderer, Renderer, {
+  /**
+   * Initialize a new HybridRenderer instance.
+   * @param {DOMElement} el - The containing DOM element for the display.
+   * @param {number} width - The coordinate width of the display, in pixels.
+   * @param {number} height - The coordinate height of the display, in pixels.
+   * @param {Array<number>} origin - The origin of the display, in pixels.
+   *   The coordinate system will be translated to this point.
+   * @param {number} [scaleFactor=1] - Optional scaleFactor by which to multiply
+   *   the width and height to determine the final pixel size.
+   * @return {HybridRenderer} - This renderer instance.
+   */
+  initialize(el, width, height, origin, scaleFactor) {
+    this._canvasEl = domChild(el, 0, 'div');
+    this._svgEl = domChild(el, 1, 'div');
+
+    // Set position absolute to overlay svg on top of canvas
+    if (!OPTS.debug) {
+      this._canvasEl.style.position = 'absolute';
+      this._svgEl.style.position = 'absolute';
+    }
+
+    // pointer-events to none on SVG layer so that canvas gets all events
+    this._svgEl.style.pointerEvents = 'none';
+
+    this._canvasRenderer.initialize(this._canvasEl, width, height, origin, scaleFactor);
+    this._svgRenderer.initialize(this._svgEl, width, height, origin, scaleFactor);
+    return base.initialize.call(this, el, width, height, origin, scaleFactor);
+  },
+
+  /**
+   * Flag a mark item as dirty.
+   * @param {Item} item - The mark item.
+   */
+  dirty(item) {
+    if (OPTS.svgMarkTypes.includes(item.mark.marktype)) {
+      this._svgRenderer.dirty(item);
+    } else {
+      this._canvasRenderer.dirty(item);
+    }
+    return this;
+  },
+
+  /**
+   * Internal rendering method.
+   * @param {object} scene - The root mark of a scenegraph to render.
+   * @param {Array} markTypes - Array of the mark types to render.
+   *                            If undefined, render all mark types
+   */
+  _render(scene, markTypes) {
+    const allMarkTypes = markTypes ?? [
+      'arc', 'area', 'image', 'line', 'path', 'rect', 'rule', 'shape', 'symbol', 'text', 'trail'
+    ];
+    const canvasMarkTypes = allMarkTypes.filter((m) => !OPTS.svgMarkTypes.includes(m));
+    this._svgRenderer.render(scene, OPTS.svgMarkTypes);
```

Comment:
I am curious about the logical impact of this setup. I'm fine with a constraint like "one gets one canvas and one svg layer," but I think the crux of the design work will be ironing out this constraint. For instance, should we have an option of which is on top? I could see one example being that text rendering occurs on an svg context beneath a canvas context that has animated content (but which should be allowed to run on top of the axis).

Reply:
Yeah, those are good points. I added support for placing the canvas layer on top in [4a44501](https://github.com/vega/vega/pull/3810/commits/4a44501051fa519e73690fef9c3ef29a0a2d6773).